src/main/java/com/sourcegraph/cody/autocomplete/CodyEditorFactoryListener.kt

```diff
@@ -0,0 +1,187 @@
+ * and document edits.
+ */
+class CodyEditorFactoryListener : EditorFactoryListener {
+  private var selectionListener = CodySelectionListener()
+  private var caretListener: CaretListener = CodyCaretListener()
+
+  override fun editorCreated(event: EditorFactoryEvent) {
+    if (!isCodyEnabled()) {
+      return
+    }
+    val editor = event.editor
+    informAgentAboutEditorChange(editor, false)
+    val project = editor.project
+    if (project == null || project.isDisposed) {
+      return
+    }
+    val disposable = Disposer.newDisposable("CodyEditorFactoryListener")
+    EditorUtil.disposeWithEditor(editor, disposable)
+    editor.caretModel.addCaretListener(caretListener, disposable)
+    editor.selectionModel.addSelectionListener(selectionListener, disposable)
+    editor.document.addDocumentListener(CodyDocumentListener(editor), disposable)
+  }
+
+  private class CodyCaretListener : CaretListener {
+    override fun caretPositionChanged(e: CaretEvent) {
+      if (!isCodyEnabled()) {
+        return
+      }
+      val commandName = CommandProcessor.getInstance().currentCommandName
+      if (commandName == VIM_EXIT_INSERT_MODE_ACTION) {
+        return
+      }
+      informAgentAboutEditorChange(e.editor)
+      val suggestions = instance
+      val editor = e.editor
+      if (isEditorValidForAutocomplete(editor) && isSelectedEditor(editor)) {
+        suggestions.clearAutocompleteSuggestions(e.editor)
+        if (isImplicitAutocompleteEnabledForEditor(editor))
+            suggestions.triggerAutocomplete(
+                e.editor, e.editor.caretModel.offset, InlineCompletionTriggerKind.AUTOMATIC)
+      }
+    }
+  }
+
+  private class CodySelectionListener : SelectionListener {
+    override fun selectionChanged(e: SelectionEvent) {
+      if (!isCodyEnabled()) {
+        return
+      }
+      val editor = e.editor
+      informAgentAboutEditorChange(editor)
+      if (isEditorValidForAutocomplete(editor) && isCodyEnabled() && isSelectedEditor(editor))
+          instance.clearAutocompleteSuggestions(editor)
+    }
+  }
+
+  private class CodyDocumentListener(private val editor: Editor) : BulkAwareDocumentListener {
+    override fun documentChangedNonBulk(event: DocumentEvent) {
+      if (!isSelectedEditor(editor)) {
+        return
+      }
+      val completions = instance
+      completions.clearAutocompleteSuggestions(editor)
+      if (isImplicitAutocompleteEnabledForEditor(editor) &&
+          isEditorValidForAutocomplete(editor) &&
+          !CommandProcessor.getInstance().isUndoTransparentActionInProgress) {
+        informAgentAboutEditorChange(editor)
+        val changeOffset = event.offset + event.newLength
+        if (editor.caretModel.offset == changeOffset) {
+          val requestType =
+              if (event.oldLength != event.newLength) InlineAutocompleteTriggerKind.Invoke
+              else InlineAutocompleteTriggerKind.Automatic
```

Comment:
@olafurpg, can you take a look at this part?

`requestType` is not used. I think it is safe to delete it.

Reply:
Fine to remove it if it's not used. We hardcode `AUTOMATIC` below, which makes sense since this event is triggered on document changes, not the alt-backslash shortcut.